### ISSUE TITLE
git checkout alias completion error: "unary operator expected"

### ISSUE THREAD

**Issue by petobens** [scop/bash-completion#545]
## Describe the bug

When using the following completion alias:
```bash
__git_complete gco _git_checkout
```
I get 
```
gco bash: [: -lt: unary operator expected
bash: [: : integer expression expected
bash: [: -lt: unary operator expected
bash: [: -lt: unary operator expected
```

## To reproduce

1. Write a sample `git_bashrc.sh` file with:
```bash
. "/usr/share/bash-completion/completions/git"
__git_complete gco _git_checkout
```
2. As in the GIF, start bash within a git dir with `bash --rcfile <path to git_bashrc.sh>`
3. Write `gco` and press `TAB`
4. See the `unary operator expected` issue mentioned above.
5. If I instead use `git checkout` (i.e avoid the alias) completion works fine

## Expected behavior

I expect completion of `gco` to work the way that `git checkout` does.

## Versions (please complete the following information)

- [ ] Operating system name/distribution and version: Arch LInux 5.12.11
- [ ] bash version, `echo "$BASH_VERSION"`: 5.18
- [ ] bash-completion version, `(IFS=.; echo "${BASH_COMPLETION_VERSINFO[*]}")`: 2.11

## Additional context

![Peek 2021-06-19 17-14](https://user-images.githubusercontent.com/2583971/122654452-d3a71300-d121-11eb-90ff-e1a30a741e17.gif)

**Comment by akinomyoga** [comment]
`/usr/share/bash-completion/completions/git` is not a part of this project. The completion file for `git` is maintained in [the repository of Git](https://github.com/git/git/blob/master/contrib/completion/git-completion.bash). I confirmed with the latest version of `git-completion.bash`. Here's the trace:

```diff
+ __git_func_wrap _git_checkout
+ local cur words cword prev
+ _get_comp_words_by_ref -n =: cur words cword prev
+ local exclude cur_ words_ cword_
+ '[' -n = -n ']'
+ exclude==:
+ shift 2
+ __git_reassemble_comp_words_by_ref =:
+ local exclude i j first
+ exclude==:
+ cword_=1
+ '[' -z =: ']'
+ (( i=0, j=0 ))
+ (( i < 2 ))
+ first=t
+ '[' 0 -gt 0 ']'
+ words_[$j]=gco
+ '[' 0 = 1 ']'
+ (( i++, j++ ))
+ (( i < 2 ))
+ first=t
+ '[' 1 -gt 0 ']'
+ '[' -n '' ']'
+ words_[$j]=
+ '[' 1 = 1 ']'
+ cword_=1
+ (( i++, j++ ))
+ (( i < 2 ))
+ cur_=
+ '[' 4 -gt 0 ']'
+ case "$1" in
+ cur=
+ shift
+ '[' 3 -gt 0 ']'
+ case "$1" in
+ words=("${words_[@]}")
+ shift
+ '[' 2 -gt 0 ']'
+ case "$1" in
+ cword=1
+ shift
+ '[' 1 -gt 0 ']'
+ case "$1" in
+ prev=gco
+ shift
+ '[' 0 -gt 0 ']'
+ _git_checkout
+ __git_has_doubledash
+ local c=1
+ '[' 1 -lt 1 ']'
+ return 1
++ __git_checkout_default_dwim_mode
++ local last_option dwim_opt=--dwim
++ '[' '' = 1 ']'
+++ __git_find_on_cmdline --no-track
+++ local word c= show_idx
+++ test 1 -gt 1
+++ local wordlist=--no-track
+++ '[' -lt 1 ']'
bash: [: -lt: unary operator expected
++ '[' -n '' ']'
+++ __git config --type=bool checkout.guess
+++ git config --type=bool checkout.guess
++ '[' '' = false ']'
+++ __git_find_last_on_cmdline '--guess --no-guess'
+++ local word c=1 show_idx
+++ test 1 -gt 1
+++ local 'wordlist=--guess --no-guess'
+++ '[' 1 -gt '' ']'
bash: [: : integer expression expected
++ last_option=
++ case "$last_option" in
++ echo --dwim
+ local dwim_opt=--dwim
+ case "$prev" in
+ case "$cur" in
++ __git_find_on_cmdline '-b -B -d --detach --orphan'
++ local word c= show_idx
++ test 1 -gt 1
++ local 'wordlist=-b -B -d --detach --orphan'
++ '[' -lt 1 ']'
bash: [: -lt: unary operator expected
+ '[' -n '' ']'
++ __git_find_on_cmdline --track
++ local word c= show_idx
++ test 1 -gt 1
++ local wordlist=--track
++ '[' -lt 1 ']'
bash: [: -lt: unary operator expected
+ '[' -n '' ']'
+ __git_complete_refs --dwim --mode=refs
+ local remote= dwim= pfx= cur_= 'sfx= ' mode=refs
+ test 2 '!=' 0
+ case "$1" in
+ dwim=yes
+ shift
+ test 1 '!=' 0
+ case "$1" in
+ mode=refs
+ shift
+ test 0 '!=' 0
+ case "$mode" in
++ __git_refs '' '' '' '' ' '
++ local i hash dir track=
++ local list_refs_from=path remote=
++ local format refs
++ local pfx= cur_= 'sfx= '
++ local match=
++ local fer_pfx=
++ __git_find_repo_path
++ '[' -n '' ']'
++ '[' -n '' ']'
++ '[' -n '' ']'
++ '[' -n '' ']'
++ '[' -d .git ']'
++ __git_repo_path=.git
++ dir=.git
++ '[' -z '' ']'
++ '[' -z .git ']'
++ '[' path = path ']'
++ [[ '' == ^* ]]
++ case "$cur_" in
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/HEAD ']'
++ echo 'HEAD '
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/FETCH_HEAD ']'
++ echo 'FETCH_HEAD '
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/ORIG_HEAD ']'
++ echo 'ORIG_HEAD '
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/MERGE_HEAD ']'
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/REBASE_HEAD ']'
++ for i in HEAD FETCH_HEAD ORIG_HEAD MERGE_HEAD REBASE_HEAD CHERRY_PICK_HEAD
++ case "$i" in
++ '[' -e .git/CHERRY_PICK_HEAD ']'
++ format=refname:strip=2
++ refs=("refs/tags/$match*" "refs/tags/$match*/**" "refs/heads/$match*" "refs/heads/$match*/**" "refs/remotes/$match*" "refs/remotes/$match*/**")
++ __git_dir=.git
++ __git for-each-ref '--format=%(refname:strip=2) ' 'refs/tags/*' 'refs/tags/*/**' 'refs/heads/*' 'refs/heads/*/**' 'refs/remotes/*' 'refs/remotes/*/**'
++ git --git-dir=.git for-each-ref '--format=%(refname:strip=2) ' 'refs/tags/*' 'refs/tags/*/**' 'refs/heads/*' 'refs/heads/*/**' 'refs/remotes/*' 'refs/remotes/*/**'
++ '[' -n '' ']'
++ return
+ __gitcomp_direct 'HEAD
FETCH_HEAD
ORIG_HEAD
[...] '
+ local 'IFS=
'
+ COMPREPLY=($1)
+ '[' yes = yes ']'
++ __git_dwim_remote_heads '' '' ' '
++ local pfx= cur_= 'sfx= '
++ local fer_pfx=
++ __git for-each-ref '--format=%(refname:strip=3) ' --sort=refname:strip=3 'refs/remotes/*/*' 'refs/remotes/*/*/**'
++ git for-each-ref '--format=%(refname:strip=3) ' --sort=refname:strip=3 'refs/remotes/*/*' 'refs/remotes/*/*/**'
++ uniq -u
+ __gitcomp_direct_append 'HEAD
[...] '
+ local 'IFS=
'
+ COMPREPLY+=($1)
```

[This](https://github.com/git/git/blob/master/contrib/completion/git-completion.bash#L1190) is caused by [an empty variable `__git_cmd_idx`](https://github.com/git/git/blob/master/contrib/completion/git-completion.bash#L1179). Normally, [`__git_cmd_idx` is set in `__git_main`](https://github.com/git/git/blob/master/contrib/completion/git-completion.bash#L3420). However, if one uses `__git_complete xxx _git_yyyy`, `__git_cmd_idx` is not set.

**Comment by petobens** [comment]
@akinomyoga thanks for looking into this. Mmm is there a workaround?  Otherwise I can do a PR on the git repo if you give me any pointers on how to fix it.

**Comment by akinomyoga** [comment]
> Mmm is there a workaround?

You can set up the value of `__git_cmd_idx` when `_git_checkout` is called:

```bash
# bashrc

source /usr/share/bash-completion/completions/git
_comp_gco() {
  local __git_cmd_idx=0
  _git_checkout
} &&
  __git_complete gco _comp_gco
```

> Otherwise I can do a PR on the git repo if you give me any pointers on how to fix it.

I think there is no general automatic solution because it seems difficult to automatically find the proper value for `__git_cmd_idx`. So user needs to manually set `__git_cmd_idx`. But I guess `__git_cmd_idx=0` would be the good default value. Maybe we could support an option of `__git_complete` to specify the custom value of `__git_cmd_idx` though this is not so user-friendly.

**Comment by petobens** [comment]
Thanks for all the help! Feel free to close this one if you deem fit.

**Comment by JBallin** [comment]
Looks like this will be fixed in a future version of git: https://github.com/git/git/pull/1038

**Comment by petobens** [comment]
Thanks! Let's close this one then.

**Comment by phil-blain** [comment]
Just noting that git/git#1038 has been closed by its author before even being sent to the Git mailing list. So the fix is not going to be in a released version of Git anytime soon.

**Comment by akinomyoga** [comment]
OK, so the fix is already applied to the development version of git? Anyway, thank you for trying to inform us of the latest situation!

Refs (where it seems people is saying that it is already fixed):

- https://github.com/gitgitgadget/git/issues/1064
- [scripts - Bash completion for git branches broken after upgrading to 21.10 - Ask Ubuntu](https://askubuntu.com/questions/1371505/bash-completion-for-git-branches-broken-after-upgrading-to-21-10)

**Comment by phil-blain** [comment]
The fix was merged before Git 2.33.0, which was released in August. So it's already in 2 released versions of Git :) I've noted that in https://github.com/gitgitgadget/git/issues/1064 and will suggest an edit to the SO question.

**Comment by akinomyoga** [comment]
OK! Great!
